src/box/lua/schema.lua

```diff
@@ -3453,35 +3453,64 @@ local function grant(uid, name, privilege, object_type,
     else
         options.grantor = user_or_role_resolve(options.grantor)
     end
+    if options.origin == nil then
+        options.origin = ''
+    elseif type(options.origin) ~= 'string' or #options.origin == 0 then
+        box.error(box.error.ILLEGAL_PARAMS, "options parameter 'origin' " ..
+                  "should be of type 'string' and should be non-empty string")
+    end
     local _priv = box.space[box.schema.PRIV_ID]
     local _vpriv = box.space[box.schema.VPRIV_ID]
     -- add the granted privilege to the current set
     local tuple = _vpriv:get{uid, object_type, oid}
-    local old_privilege
-    if tuple ~= nil then
-        old_privilege = tuple.privilege
+
+    local origins
+    if tuple == nil then
+        origins = {[''] = 0}
+    elseif tuple[6] == nil or tuple[6].origins == nil then
+        origins = {[''] = tuple.privilege}
     else
-        old_privilege = 0
+        origins = tuple[6].origins
     end
-    privilege_hex = bit.bor(privilege_hex, old_privilege)
-    privilege_hex = tonumber(ffi.cast('uint32_t', privilege_hex))
+
     -- do not execute a replace if it does not change anything
     -- XXX bug if we decide to add a grant option: new grantor
     -- replaces the old one, old grantor is lost
-    if privilege_hex ~= old_privilege then
-        _priv:replace{options.grantor, uid, object_type, oid, privilege_hex}
-    elseif not options.if_not_exists then
-            if object_type == 'role' and object_name ~= '' and
-               privilege == 'execute' then
-                box.error(box.error.ROLE_GRANTED, name, object_name)
-            else
-                if not is_singleton_object_type(object_type) then
-                    object_name = string.format(" '%s'", object_name)
-                end
-                box.error(box.error.PRIV_GRANTED, name, privilege,
-                          object_type, object_name)
-            end
+    local old_privilege = origins[options.origin] or 0
+    if bit.band(bit.bnot(old_privilege), privilege_hex) ~= 0 then
+        -- Update given privileges by origin.
+        origins[options.origin] = bit.bor(old_privilege, privilege_hex)
+    else
+        -- No new privileges can be given.
+        if options.if_not_exists then
+            return
+        end
+        if object_type == 'role' and object_name ~= '' and
+           privilege == 'execute' then
+            box.error(box.error.ROLE_GRANTED, name, object_name)
+        end
+        if not is_singleton_object_type(object_type) then
+            object_name = string.format(" '%s'", object_name)
+        end
+        box.error(box.error.PRIV_GRANTED, name, privilege, object_type,
+                  object_name)
```

Comment:
The error message now can be confusing. Let's look on your example from the documentation request.

```lua
box.schema.role.create('a')
box.schema.role.grant('a', 'read, write', 'space', '_space', {origin = 'one'})
box.schema.role.grant('a', 'read', 'space', '_space', {origin = 'two'})
box.schema.role.revoke('a', 'write', 'space', '_space', {origin = 'two'})
-- error: User 'a' does not have write access on space '_space'
```

The role 'a' has the `write` privilege on the space `_space`. It has no `write` privilege on this space granted by `two`.

I would also appreciate if this information will be presented in `box.schema.{user,role}.info()`. Otherwise it may be hard to debug what is going on. However, I'm OK if the `.info()` adjustment will be implemented in a separate patchset.

Reply:
It was decided to not make this feature public, so it will not be added to `info()`.

Reply:
I still think that the error may be very confusing: there is a privilege, it works, but an attempt to revoke it says that there is no such a privilege.

We discussed it with Mergen in a chat and we agreed on the following solution.

1. A check against `tuple.prigileve` is performed first. It issues an error with the old message if there is no needed privilege.
2. Then, a check against a given `tuple[6].origins[opts._origin]` is performed. If there is no needed privilege, an error says that it is missed *in the given origin*.

Reply:
The new error messages were added. Thanks!